bouncer/shared/broker_fee_collection.ts

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/env -S pnpm tsx
+import assert from 'assert';
+import { randomBytes } from 'crypto';
+import Keyring from '@polkadot/keyring';
+import { Asset, Assets } from '@chainflip/cli';
+import { cryptoWaitReady } from '@polkadot/util-crypto';
+import {
+  EgressId,
+  amountToFineAmount,
+  brokerMutex,
+  decodeDotAddressForContract,
+  handleSubstrateError,
+  newAddress,
+  observeBalanceIncrease,
+  observeEvent,
+  shortChainFromAsset,
+  assetDecimals,
+  hexStringToBytesArray,
+  getChainflipApi,
+} from '../shared/utils';
+import { getBalance } from '../shared/get_balance';
+import { doPerformSwap } from '../shared/perform_swap';
+
+const swapAssetAmount = {
+  [Assets.ETH]: 1,
+  [Assets.DOT]: 1000,
+  [Assets.FLIP]: 1000,
+  [Assets.BTC]: 0.1,
+  [Assets.USDC]: 1000,
+};
+const commissionBps = 1000; // 10%
+
+const chainflip = await getChainflipApi();
+
+const keyring = new Keyring({ type: 'sr25519' });
+const broker = keyring.createFromUri('//BROKER_FEE_TEST');
+
+export async function submitBrokerWithdrawal(
+  asset: Asset,
+  addressObject: { [chain: string]: string },
+) {
+  // Only allow one withdrawal at a time to stop nonce issues
+  return brokerMutex.runExclusive(async () =>
+    chainflip.tx.swapping
+      .withdraw(asset, addressObject)
+      .signAndSend(broker, { nonce: -1 }, handleSubstrateError(chainflip)),
+  );
+}
+
+/// Runs a swap, checks that the broker fees are collected,
+/// then withdraws the broker fees, making sure the balance is correct after the withdrawal.
+async function testBrokerFees(asset: Asset, seed?: string): Promise<void> {
+  // Check the broker fees before the swap
+  const earnedBrokerFeesBefore = BigInt(
+    (await chainflip.query.swapping.earnedBrokerFees(broker.address, asset)).toString(),
+  );
+  console.log(`${asset} earnedBrokerFeesBefore:`, earnedBrokerFeesBefore);
+
+  // Run a swap
+  const swapAsset = asset === Assets.USDC ? Assets.FLIP : Assets.USDC;
+  const destinationAddress = await newAddress(swapAsset, seed ?? randomBytes(32).toString('hex'));
+  const observeDestinationAddress =
+    asset === Assets.DOT ? decodeDotAddressForContract(destinationAddress) : destinationAddress;
+  const destinationChain = shortChainFromAsset(swapAsset); // "ETH" -> "Eth"
+  console.log(`${asset} destinationAddress:`, destinationAddress);
+  const observeSwapScheduledEvent = observeEvent(
+    ':SwapScheduled',
+    chainflip,
+    (event) =>
+      event.data.destinationAddress[destinationChain]?.toLowerCase() ===
+      observeDestinationAddress.toLowerCase(),
+  );
+  console.log(`Running swap ${asset} -> ${swapAsset}`);
+
+  const rawDepositForSwapAmount = swapAssetAmount[asset].toString();
+
+  // we need to manually create the swap channel and observe the relative event
+  // because we want to use a separate broker to not interfere with other tests
```

Comment:
Could be worth changing `performSwap` to have an optional argument for the broker URL, then we can avoid this duplication.